### PR TITLE
[WIP] Implement ADD opcode.

### LINKED ISSUE #9: [EVM] `ADD` example
Implement the `ADD` opcode in the EVM circuit.

This requires a bus mapping as input.

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #9. Based on #21.

Documented at https://hackmd.io/@liangcc/zkvmbook/%2Fq8EhMIp_TimpPWWFXtwOVw.

- [x] `global_counter` checks
- [x] `stack_pointer` checks
- [ ] bus mapping lookups for reads and write

### CODE REVIEW

File: src/evm_circuit/arithmetic/add.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                    sp_second_read - sp_write
+                };
+
+                array::IntoIter::new([
+                    ("first_sp_check", first_sp_check),
+                    ("second_sp_check", second_sp_check),
+                ])
+            };
+
+            // TODO: Lookup both reads in bus mapping
```
Comment (barryWhiteHat) [inline src/evm_circuit/arithmetic/add.rs:98]: We got to lookup the write after the addition to make sure its put on the stack  ?

Comment (therealyingtong) [reply src/evm_circuit/arithmetic/add.rs:98]: Is this covered by my `TODO` on L99?
```
// TODO: Lookup write in bus mapping
```

Comment (barryWhiteHat) [reply src/evm_circuit/arithmetic/add.rs:98]: Yes think i missed that :/

**Comment by therealyingtong** [comment]
This has to be rebased on main.

File: src/evm_circuit/arithmetic/add.rs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        c: [Option<u8>; 32],
+        carry: [Option<bool>; 32],
+    ) -> Result<(Word<F>, Word<F>, Word<F>), Error> {
+        #[cfg(test)]
+        for idx in 0..31 {
+            let a = a[idx].unwrap_or(0);
+            let b = b[idx].unwrap_or(0);
+            let c = c[idx].unwrap_or(0);
+            let carry_in = carry[idx].unwrap_or(false);
+            let carry_out = carry[idx + 1].unwrap_or(false);
```
Comment (han0110) [inline src/evm_circuit/arithmetic/add.rs:191]: `idx = 31` assertion is missing, would checking like this way be more expressive?

```suggestion
        for idx in 0..32 {
            let a = a[idx].unwrap_or(0);
            let b = b[idx].unwrap_or(0);
            let c = c[idx].unwrap_or(0);
            let carry_in = if idx == 0 { None } else { carry[idx] }.unwrap_or(false);
            let carry_out = if idx == 31 { carry[0] } else { carry[idx + 1] }.unwrap_or(false);
```

File: src/evm_circuit/arithmetic/add.rs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+                    |mut region| config.assign(&mut region, 0, self.a, self.b, self.c, self.carry),
+                )?;
+
+                Ok(())
+            }
+        }
+
+        let a = pallas::Base::rand().to_bytes();
+        let b = pallas::Base::rand().to_bytes();
+        let mut c: Vec<u8> = Vec::with_capacity(32);
```
Comment (han0110) [inline src/evm_circuit/arithmetic/add.rs:301]: unnecessary annotation

```suggestion
        let mut c = Vec::with_capacity(32);
```

**Comment by CPerezz** [comment]
This will need to be rebased into `main` now that https://github.com/appliedzkp/zkevm-circuits/pull/41 was merged.

**Comment by therealyingtong** [comment]
Closed in favour of #26

**Pull request closed without merging**